### PR TITLE
bf16 AMP + fused AdamW for flow and confidence trainers

### PR DESCRIPTION
**Pull request by vratins**
Adding mixed precision (bfloat16) toggles in all training scripts.

## Summary by CodeRabbit

* **New Features**
  * Added optional CUDA bfloat16 automatic mixed precision for training and validation, enabled by default.
  * Added an option to enable fused AdamW optimization on CUDA-compatible hardware.
  * Added command-line controls to disable mixed precision when needed.

* **Bug Fixes**
  * Training automatically falls back to standard precision and optimizer behavior on non-CUDA devices.
  * Loss calculations remain stable by using full precision for final comparisons.

### CODE REVIEW

File: src/flow.py [hunk 1]
```diff
@@ -1122,16 +1134,17 @@ def training_step(
...
+            v_pred = self.model(batch, t)
 
         # target velocity
         v_target = x1_star - x0_star
 
-        # weighted MSE loss (upweight near t=1)
+        # weighted MSE loss (upweight near t=1), reduced in fp32
         w = 1.0 / (self.loss_eps + (1.0 - t_per_atom))
-        per_atom_mse = (v_pred - v_target).pow(2).mean(dim=-1, keepdim=True)
+        per_atom_mse = (v_pred.float() - v_target).pow(2).mean(dim=-1, keepdim=True)
```
Comment (DorisMai) [inline src/flow.py]: would you need to cast it again for the rmsd calculation below? and same thing in validation (line 1220-1224)

File: scripts/train.py [hunk 2]
```diff
@@ -1130,6 +1144,9 @@ def main():
     else:
         device = torch.device(args.device if torch.cuda.is_available() else "cpu")
 
+    if args.use_amp and device.type != "cuda":
```
Comment (DorisMai) [inline scripts/train.py]: why not set `args.use_amp=False` here? does config json save this? same in train_confidency.py

File: scripts/train_confidence.py [hunk 3]
```diff
@@ -470,6 +482,7 @@ def train_one_epoch(
     total_loss, total_n = 0.0, 0
     params = [p for group in optimizer.param_groups for p in group["params"]]
     accum = max(1, args.grad_accum_steps)
+    use_amp = getattr(args, "use_amp", True) and device.type == "cuda"
```
Comment (DorisMai) [inline scripts/train_confidence.py]: why not use `args.use_amp`? same in line 574

**Review by DorisMai — APPROVED** [review]
I don't have direct experience with amp to review this properly, but try to look at coderabbitai's comments

File: scripts/train_confidence.py [hunk 4]
```diff
@@ -565,8 +582,9 @@ def validate(
         target = batch["water"].target_confidence
         if target.numel() == 0:
             continue
-        preds = model(batch, return_logits=True)
-        loss = F.binary_cross_entropy_with_logits(preds, target)
+        with torch.autocast(device_type="cuda", dtype=torch.bfloat16, enabled=use_amp):
+            preds = model(batch, return_logits=True)
+            loss = F.binary_cross_entropy_with_logits(preds, target)
```
[the remark this answers is not in the document]
Comment (vratins) [reply scripts/train_confidence.py:587]: Yep makes sense, casting once right after autocast

File: scripts/train.py (hunk 2 above)
Comment (vratins) [reply scripts/train.py]: That's fair, fixed it for both train scripts.

File: scripts/train_confidence.py (hunk 3 above)
Comment (vratins) [reply scripts/train_confidence.py]: Done

File: src/flow.py [hunk 5]
```diff
@@ -935,6 +939,14 @@ def __init__(
             self.graph_cutoff = DEFAULT_EDGE_CUTOFF
         self.sampling_strategy = sampling_strategy
 
+    def _autocast_context(self, device: torch.device):
+        """bf16 autocast on CUDA when --use_amp is set; a no-op otherwise."""
+        return torch.autocast(
+            device_type="cuda",
+            dtype=torch.bfloat16,
+            enabled=self.use_amp and device.type == "cuda",
+        )
```
[the remark this answers is not in the document]
Comment (vratins) [reply src/flow.py:948]: Added a guard.

File: src/flow.py (hunk 1 above)
Comment (vratins) [reply src/flow.py]: Yep fair, casting it once and re-using.

Comment (vratins) [reply src/flow.py]: As far as the review is concerned, torch does abstract a lot of the amp usage when using bfloat16 in the latest versions, so less likely to error out compared to when we needed context managers.

File: src/flow.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (vratins) [reply src/flow.py:948]: Check again

[the remark this answers is not in the document]
Comment (vratins) [reply src/flow.py:948]: done
